plotter/conrec.go

```diff
@@ -0,0 +1,177 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package plotter
+
+import (
+	"math"
+)
+
+type point struct {
+	X, Y float64
+}
+
+type line struct {
+	p1, p2 point
+}
+
+func sect(h, p [5]float64, v1, v2 int) float64 {
+	return (h[v2]*p[v1] - h[v1]*p[v2]) / (h[v2] - h[v1])
+}
+
+// conrecLine performs an operation with a line at a given height derived
+// from data over the 2D box interval (i, j) to (i+1, j+1).
+type conrecLine func(i, j int, l line, height float64)
+
+// conrec is a Go translation of the C version of CONREC by Paul Bourke:
+// http://paulbourke.net/papers/conrec/conrec.c
+//
+// conrec takes g, an m×n grid function, a sorted slice of contour heights
+// and a conrecLine function.
+//
+// For full details of the algorithm, see the paper at
+// http://paulbourke.net/papers/conrec/
+func conrec(g GridXYZ, heights []float64, fn conrecLine) {
+	var (
+		p1, p2 point
+
+		h      [5]float64
+		sh     [5]int
+		xh, yh [5]float64
+
+		im = [4]int{0, 1, 1, 0}
+		jm = [4]int{0, 0, 1, 1}
+
+		cases = [3][3][3]int{
+			{{0, 0, 8}, {0, 2, 5}, {7, 6, 9}},
+			{{0, 3, 4}, {1, 0, 1}, {4, 3, 0}},
```

Comment:
Yes. I wrote to Paul about that, but haven't had a reply.

I believe the 3 in that position is incorrect when considering the commentary in the paper.

axiom: When we have a section of the grid where all the Z values are equal, and equal to a contour height we would expect to have no internal segments to draw.

This is covered by case g) in his description of the algorithm (a triangle with three vertices the lie on the contour level). He says "... case g above has no really satisfactory solution and fortunately will occur rarely with real arithmetic." and then goes on to show the following:

![conrec cases](http://paulbourke.net/papers/conrec/conrec3.gif)

which shows case g) in the set where no edge is drawn.

However, in the iteration over sh at conrec.c +44 (conrec.go +74), a triangle with all vertices on the plane is given sh = {0,0,0,0,0} and then when the switch happens, castab/cases resolves that to case 3 for all values of m. This is fixed by replacing castab/cases[1][1][1] with 0.